libraries/hotstuff/test/hotstuff_tools.cpp

```diff
@@ -0,0 +1,63 @@
+#include <boost/test/unit_test.hpp>
+#include <eosio/chain/types.hpp>
+#include <eosio/chain/block_header.hpp>
+
+#include <fc/exception/exception.hpp>
+
+#include <eosio/hotstuff/qc_chain.hpp>
+
+BOOST_AUTO_TEST_CASE(view_number_tests) try {
+
+  eosio::hotstuff::hs_proposal_message hspm_1;
+  eosio::hotstuff::hs_proposal_message hspm_2;
+  eosio::hotstuff::hs_proposal_message hspm_3;
+  eosio::hotstuff::hs_proposal_message hspm_4;
+  eosio::hotstuff::hs_proposal_message hspm_5;
+
+  hspm_1.block_id = eosio::chain::block_id_type("0b93846ba73bdfdc9b2383863b64f8f921c8a2379d6dde4e05bdd2e434e9392a"); //UX Network block #194217067
+  hspm_1.phase_counter = 0;
+
+  hspm_2.block_id = eosio::chain::block_id_type("0b93846ba73bdfdc9b2383863b64f8f921c8a2379d6dde4e05bdd2e434e9392a"); //UX Network block #194217067
+  hspm_2.phase_counter = 1;
+
+  hspm_3.block_id = eosio::chain::block_id_type("0b93846cf55a3ecbcd8f9bd86866b1aecc2e8bd981e40c92609ce3a68dbd0824"); //UX Network block #194217068
+  hspm_3.phase_counter = 0;
+
+  hspm_4.block_id = eosio::chain::block_id_type("0b93846cf55a3ecbcd8f9bd86866b1aecc2e8bd981e40c92609ce3a68dbd0824"); //UX Network block #194217068
+  hspm_4.phase_counter = 1;
+
+  hspm_5.block_id = eosio::chain::block_id_type("0b93846cf55a3ecbcd8f9bd86866b1aecc2e8bd981e40c92609ce3a68dbd0824"); //UX Network block #194217068
+  hspm_5.phase_counter = 2;
+
+  eosio::hotstuff::view_number vn_1 = hspm_1.get_view_number();
+  eosio::hotstuff::view_number vn_2 = hspm_2.get_view_number();
+  eosio::hotstuff::view_number vn_3 = hspm_3.get_view_number();
+  eosio::hotstuff::view_number vn_4 = hspm_4.get_view_number();
+  eosio::hotstuff::view_number vn_5 = hspm_5.get_view_number();
+
+  //test getters
+
+  BOOST_CHECK_EQUAL(vn_1.block_height() == 194217067, true);
+  BOOST_CHECK_EQUAL(vn_1.phase_counter() == 0, true);
+
+  //test operators == true
+  BOOST_CHECK_EQUAL(vn_1 == vn_1, true);
+  BOOST_CHECK_EQUAL(vn_1 != vn_2, true);
+
+  BOOST_CHECK_EQUAL(vn_1 < vn_2, true);
```

Comment:
Use `BOOST_CHECK_LT`

Reply:
These macros (LE, LE, GE, GT, etc.) are not working here for some reason (producing dozens of pages of template errors). I'm simply using `BOOST_CHECK` which at least eliminates the `true`.

Reply:
> These macros (LE, LT, GE, GT, etc.) are not working here for some reason (producing dozens of pages of template errors). I'm simply using `BOOST_CHECK` which at least eliminates the `true`.

That's because your class doesn't have an "output" function. Add this to `struct view_number`:

```c++
   struct view_number {
      // ... other functions
      friend std::ostream& operator<<(std::ostream& os, const view_number& vn) {
         os << "view_number(" << vn.bheight << ", " << vn.pcounter << ")\n";
      }
```